Introduce BorderShape to prepare for non-rounded corner shapes

smfr opened the pull request:
#### f48555f2c5d09dba8dc48ad5383ad982265c84be
<pre>
Introduce BorderShape to prepare for non-rounded corner shapes
<a href="https://bugs.webkit.org/show_bug.cgi?id=278269">https://bugs.webkit.org/show_bug.cgi?id=278269</a>
<a href="https://rdar.apple.com/134102043">rdar://134102043</a>

Reviewed by Antti Koivisto.

There&apos;s a lot of code currently that assumes that CSS boxes are rects or rounded rects.
With `corner-shape`, that&apos;s no longer true, so we need a class that hides the complexities
of clipping, painting and hit-testing more complex border shapes.

This change introduces the `BorderShape` class, and uses it in all the places that used
to call `RenderStyle::getRoundedBorderFor()` which can be removed. Many call sites
continue to use RoundedRects; these will gradually be migrated to a form where BorderShape
itself handles the functionality, but for now, call `BorderShape::deprecated...`
functions to keep code working.

`RenderBox::borderRadii()` had some surprising code that actually returns a rounded
rect for the padding box; noted with a FIXME. This is only called from WebKitLegacy
and WebKit.

A future change will adopt BorderShape for call sites of `RenderStyle::getRoundedInnerBorderFor()`.

* LayoutTests/interaction-region/border-radii-expected.txt: The event region rects are now pixel-snapped.
* LayoutTests/interaction-region/position-only-update-expected.txt: Ditto
* LayoutTests/tiled-drawing/scrolling/non-fast-region/wheel-handler-in-columns-expected.txt: Ditto
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
* Source/WebCore/platform/graphics/RoundedRect.cpp:
(WebCore::RoundedRect::Radii::areRenderableInRect const):
(WebCore::RoundedRect::Radii::makeRenderableInRect):
(WebCore::RoundedRect::isRenderable const):
(WebCore::RoundedRect::adjustRadii):
* Source/WebCore/platform/graphics/RoundedRect.h:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::backgroundRoundedRect const):
(WebCore::BackgroundPainter::paintBoxShadow const):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::pathForBorderArea):
(WebCore::BorderPainter::paintBorder const):
(WebCore::BorderPainter::borderRectAdjustedForBleedAvoidance const):
(WebCore::BorderPainter::borderInnerRectAdjustedForBleedAvoidance const): Deleted.
* Source/WebCore/rendering/BorderPainter.h:
* Source/WebCore/rendering/BorderShape.cpp: Added.
(WebCore::calcRadiiFor):
(WebCore::BorderShape::shapeForBorderRect):
(WebCore::BorderShape::BorderShape):
(WebCore::BorderShape::deprecatedRoundedRect const):
(WebCore::BorderShape::deprecatedInnerRoundedRect const):
(WebCore::BorderShape::deprecatedPixelSnappedRoundedRect const):
(WebCore::BorderShape::deprecatedPixelSnappedInnerRoundedRect const):
(WebCore::BorderShape::snappedOuterRect const):
(WebCore::BorderShape::snappedInnerRect const):
(WebCore::addRoundedRectToPath):
(WebCore::BorderShape::pathForOuterShape const):
(WebCore::BorderShape::pathForInnerShape const):
(WebCore::BorderShape::pathForBorderArea const):
(WebCore::BorderShape::clipToOuterShape):
(WebCore::BorderShape::clipToInnerShape):
(WebCore::BorderShape::clipOutInnerShape):
(WebCore::BorderShape::fillOuterShape):
(WebCore::BorderShape::innerEdgeRoundedRect const):
(WebCore::BorderShape::innerEdgeRect const):
* Source/WebCore/rendering/BorderShape.h: Added.
(WebCore::BorderShape::radii const):
(WebCore::BorderShape::setRadii):
(WebCore::BorderShape::isRounded const):
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::containingBlockRectForRenderer):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::paintObject):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::borderRadii const):
(WebCore::RenderBox::hitTestBorderRadius const):
(WebCore::RenderBox::paintBoxDecorations):
(WebCore::RenderBox::borderRoundedRect const): Deleted.
(WebCore::RenderBox::roundedBorderBoxRect const): Deleted.
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintAfterLayoutIfNeeded):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateBackdropFiltersGeometry):
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::paintBoxDecorations):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::paint):
* Source/WebCore/rendering/ios/RenderThemeIOS.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::paintTextFieldDecorations):
(WebCore::RenderThemeIOS::addRoundedBorderClip): Deleted. It was unused.
* Source/WebCore/rendering/shapes/BoxShape.cpp:
(WebCore::computeRoundedRectForBoxShape):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::getRoundedBorderFor const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: <a href="https://commits.webkit.org/282528@main">https://commits.webkit.org/282528@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af280fb7c8ffaf6e8afd5e2dc4df9cb1a1588f12

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63391 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42747 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15988 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67412 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13999 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65511 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50435 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14279 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9681 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66460 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39687 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54898 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31750 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36368 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12250 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12871 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/57905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12576 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69108 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7338 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/12173 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/58373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7370 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54970 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14046 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6122 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/38568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39647 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40759 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39390 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -564,9 +565,8 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
-            path.addRoundedRect(roundedRect);
-            clipPath = path;
+            clipPath = borderShape.pathForOuterShape(renderBox->document().deviceScaleFactor());
+            WTF_ALWAYS_LOG("interactionRegionForRenderedRegion - rounded rect" << borderShape.deprecatedRoundedRect() << " device scale " << renderBox->document().deviceScaleFactor() << " " << clipPath);
```
This should log to a channel

smfr replied:
I'm trying to debug the VisionOS failure.

smfr committed:
Introduce BorderShape to prepare for non-rounded corner shapes
https://bugs.webkit.org/show_bug.cgi?id=278269
rdar://134102043

Reviewed by Antti Koivisto.

There's a lot of code currently that assumes that CSS boxes are rects or rounded rects.
With `corner-shape`, that's no longer true, so we need a class that hides the complexities
of clipping, painting and hit-testing more complex border shapes.

This change introduces the `BorderShape` class, and uses it in all the places that used
to call `RenderStyle::getRoundedBorderFor()` which can be removed. Many call sites
continue to use RoundedRects; these will gradually be migrated to a form where BorderShape
itself handles the functionality, but for now, call `BorderShape::deprecated...`
functions to keep code working.

`RenderBox::borderRadii()` had some surprising code that actually returns a rounded
rect for the padding box; noted with a FIXME. This is only called from WebKitLegacy
and WebKit.

A future change will adopt BorderShape for call sites of `RenderStyle::getRoundedInnerBorderFor()`.

* LayoutTests/interaction-region/border-radii-expected.txt: The event region rects are now pixel-snapped.
* LayoutTests/interaction-region/position-only-update-expected.txt: Ditto
* LayoutTests/tiled-drawing/scrolling/non-fast-region/wheel-handler-in-columns-expected.txt: Ditto
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
* Source/WebCore/platform/graphics/RoundedRect.cpp:
(WebCore::RoundedRect::Radii::areRenderableInRect const):
(WebCore::RoundedRect::Radii::makeRenderableInRect):
(WebCore::RoundedRect::isRenderable const):
(WebCore::RoundedRect::adjustRadii):
* Source/WebCore/platform/graphics/RoundedRect.h:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::backgroundRoundedRect const):
(WebCore::BackgroundPainter::paintBoxShadow const):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::pathForBorderArea):
(WebCore::BorderPainter::paintBorder const):
(WebCore::BorderPainter::borderRectAdjustedForBleedAvoidance const):
(WebCore::BorderPainter::borderInnerRectAdjustedForBleedAvoidance const): Deleted.
* Source/WebCore/rendering/BorderPainter.h:
* Source/WebCore/rendering/BorderShape.cpp: Added.
(WebCore::calcRadiiFor):
(WebCore::BorderShape::shapeForBorderRect):
(WebCore::BorderShape::BorderShape):
(WebCore::BorderShape::deprecatedRoundedRect const):
(WebCore::BorderShape::deprecatedInnerRoundedRect const):
(WebCore::BorderShape::deprecatedPixelSnappedRoundedRect const):
(WebCore::BorderShape::deprecatedPixelSnappedInnerRoundedRect const):
(WebCore::BorderShape::snappedOuterRect const):
(WebCore::BorderShape::snappedInnerRect const):
(WebCore::addRoundedRectToPath):
(WebCore::BorderShape::pathForOuterShape const):
(WebCore::BorderShape::pathForInnerShape const):
(WebCore::BorderShape::pathForBorderArea const):
(WebCore::BorderShape::clipToOuterShape):
(WebCore::BorderShape::clipToInnerShape):
(WebCore::BorderShape::clipOutInnerShape):
(WebCore::BorderShape::fillOuterShape):
(WebCore::BorderShape::innerEdgeRoundedRect const):
(WebCore::BorderShape::innerEdgeRect const):
* Source/WebCore/rendering/BorderShape.h: Added.
(WebCore::BorderShape::radii const):
(WebCore::BorderShape::setRadii):
(WebCore::BorderShape::isRounded const):
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::containingBlockRectForRenderer):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::paintObject):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::borderRadii const):
(WebCore::RenderBox::hitTestBorderRadius const):
(WebCore::RenderBox::paintBoxDecorations):
(WebCore::RenderBox::borderRoundedRect const): Deleted.
(WebCore::RenderBox::roundedBorderBoxRect const): Deleted.
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintAfterLayoutIfNeeded):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateBackdropFiltersGeometry):
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::paintBoxDecorations):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::paint):
* Source/WebCore/rendering/ios/RenderThemeIOS.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::paintTextFieldDecorations):
(WebCore::RenderThemeIOS::addRoundedBorderClip): Deleted. It was unused.
* Source/WebCore/rendering/shapes/BoxShape.cpp:
(WebCore::computeRoundedRectForBoxShape):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::getRoundedBorderFor const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: https://commits.webkit.org/282528@main
